doc/TockBinaryFormat.md

```diff
@@ -293,6 +309,35 @@ a non-position-independent binary at an incorrect location.
     the linker. If a fixed address is not required this should be set to
     `0xFFFFFFFF`.
 
+#### `6` Permissions
+
+The `Permissions` section allows an app to specify driver permissions that it
+is allowed to use. All driver syscalls that an app will use must be listed. The
+list should not include drivers that are not being used by the app.
+
+The data is stored in the optional `TbfHeaderV2Permissions` field. This
+includes an array of all the `perms`.
```

Comment:
What is the behavior when things are not specified? The obvious backwards-compatible choice is to default to permitted, but as a security-oriented operating system, is that actually the default we want to put into the world?

---

A perhaps more nuanced version might be that if an application specifies _no_ permissions, then it has all permissions. However, once an application chooses to specify _any_ permissions, we expect it to specify all of its requested permissions?

Reply:
I think by default we would allow all with no permissions.

Eventually I think we could then migrate to a disabled by default once all the usersapce apps/tools have migrated.

Reply:
Does "disabled by default" mean disabled if the permissions header is not present, or disabled if the permissions header is present but empty?

I would oppose a disable by default if the permissions header is not present for the following reasons:

A. This is not compatible with drivers that use all 32 bits of the command ID (i.e. have command numbers >= 64)
B. Particular projects using Tock that have their own security concerns may need their own permissions header, and that would require them to carry both sets of headers. That could waste a lot of space.

Reply:
Eventually I would say "disabled by default" means both, so access is disabled if there is no permissions header or if it is empty. Like I said though, I think initially we would allow everything and only once everything is working and we have had a release or two for everything to catch up would we make that switch.

In saying that, this PR has no impact on that decision, we could just always default to enable by default and have a kernel config that disables access.

Yep, this PR is limited to less then 64 command numbers. Some drivers will need to have this numbers changed (ninedef) to meet this. As discussed on the call we allow doubling the driver numbers to meet more bit masks.

The idea is this would be the header all apps use. If others want to duplicate this we can't really do anything to help that.

Reply:
> Eventually I would say "disabled by default" means both, so access is disabled if there is no permissions header or if it is empty. Like I said though, I think initially we would allow everything and only once everything is working and we have had a release or two for everything to catch up would we make that switch.
> 
> In saying that, this PR has no impact on that decision, we could just always default to enable by default and have a kernel config that disables access.

If you think that discussion is outside the scope of this PR, then I am fine having it elsewhere.

I do want to point out that OpenTitan has not decided what process permissions model it needs, and we cannot determine now whether OpenTitan will use this permissions model. By default, I'd assume that OpenTitan will have its own permissions system, so we want to support that case.